Description: refine setcap output on errors
 The error message of setcap is a little confused. From the error message it is hard to detect what is wrong. I removed the last 'usage' function call, because at that code point the 'usage' can not provide any useful information. Instead, I added a function 'mystrerror' which shall provide more information.
Author: Zhi Li <lizhi1215@gmail.com>
Forwarded: no
Bug: http://bugs.debian.org/cgi-bin/bugreport.cgi?bug=633075

Index: libcap-2.24/progs/setcap.c
===================================================================
--- libcap-2.24.orig/progs/setcap.c	2014-01-06 03:24:48.791188475 +0000
+++ libcap-2.24/progs/setcap.c	2014-01-06 03:25:21.531121441 +0000
@@ -22,6 +22,25 @@
     exit(1);
 }
 
+static const char * mystrerror(int n)
+{
+  struct my_error {
+    int num;
+    const char *desp;
+  }db[] = {
+    { EINVAL, "The value of the capability argument is not permitted for a file. Or the file is not a regular (non-symlink) file" },
+    { ENODATA, "Are you removing capabilities from a file? That file does not have any capability."},
+    {0, ""}
+  };
+
+  struct my_error *p = &db[0];
+  while (p->num) {
+    if ( n == p->num ) return p->desp;
+    p++;
+  }
+  return "";
+}
+
 #define MAXCAP  2048
 
 static int read_caps(int quiet, const char *filename, char *buffer)
@@ -195,7 +214,8 @@
 			"Failed to set capabilities on file `%s' (%s)\n",
 			argv[0], strerror(errno));
 		if (!explained) {
-		    usage();
+			fprintf(stderr, "%s\n", mystrerror(errno));
+			exit(1);
 		}
 	    }
 	}
